### DISCUSSION TITLE
What do you think about my Math extension?

### DISCUSSION THREAD

**Discussion by GiovanniSalmeri** [datenstrom/community#721] [Ask a question] [answered]
A new extension: https://github.com/GiovanniSalmeri/yellow-math It allows to write mathematics in AsciiMath, or in TeX. The rendering engine is the very fast KaTeX. In order to allow the use of ASCIIMath (which is not supported by KaTeX) I translated from Javascript to PHP (in order to reduce the load client-side) a program which translate ASCIIMath into TeX, and added moreover a little Javascript code which avoids the full scanning of the page by KaTeX. The result is indeed very fast (even if the AsciiMath to TeX translator could be optimised in several aspects) and the use of the extension very simple. Perhaps someone will find it useful!

**Comment by markseuffert** [comment, marked as answer]
Good work, I always wanted a math extension. For a long time I have been looking out for a PHP library that can render math on the server-side, something that doesn't need a client-side JavaScript library. As far as I know this is still not possible in 2021 and most math rendering is done on the client-side via KaTeX or MathJax. 

Here are some suggestions/thoughts:

* Include third party JavaScript and CSS bundles, would make the extension self-hosted. The advantage with self-hosted websites is that you don't rely on third party web servers and content delivery networks, it even works when you work offline on your built-in web server.
* Use shortcut `[math]` and block ` ```math ` for the AsciiMath syntax,  would make it more intuitive to use.
* Use shortcut `[tex]` and block ` ```tex ` for the TeX syntax, I know it's a violation of the usual design pattern where the shortcut/block has a name that corresponds to the extensions name, but maybe worth an exception. As an alternative you could use `[mathtex]` and ` ```mathtex ` for the TeX syntax.
* Can you elaborate on your thoughts for `MathPlainCode`, isn't inline code expected to be rendered as plain text?

Hope you find this feedback useful.

**Reply by GiovanniSalmeri** [reply]
Thank you very much for your feedback!

> Include third party JavaScript and CSS bundles, would make the extension self-hosted. The advantage with self-hosted websites is that you don't rely on third party web servers and content delivery networks, it even works when you work offline on your built-in web server.

I was doing so, but I stopped when I realised that, besides JS and CSS, I would have to include *sixty* font files (fortunately not huge: 1.1 MB overall). If you see nothing in contrary, I can do this right now.

> Use shortcut [math] and block ```math for the AsciiMath syntax, would make it more intuitive to use.
> Use shortcut [tex] and block ```tex for the TeX syntax, I know it's a violation of the usual design pattern where the shortcut/block has a name that corresponds to the extensions name, but maybe worth an exception. As an alternative you could use [mathtex] and ```mathtex for the TeX syntax.

Done! I have chosen `mathtex`, as a form of light dissuasion against the use of TeX (four letters more to type :-). Indeed I have made this extension as a means to use AsciiMath, which seems to me the "markdown of mathematics", I consider the possibility of using TeX only as an added bonus.

> Can you elaborate on your thoughts for MathPlainCode, isn't inline code expected to be rendered as plain text?

Yes. But the "native" method of enclosing mathematical expressions in AsciiMath is the backtick (as in Tex the $), and I have thought that if in a specific page there is a very frequent use of math and no use of other type of code, it could be easier to add to that page a setting `MathPlainCode: math`, and avoid for example writing a hundred times `[math a]` only for citing a variable. If you think this is confusing, no problem in suppressing this option.

**Reply by GiovanniSalmeri** [reply]
As a side-note: I realised that in this extension (and probably also in other ones) I used features of PHP 7.0 (for example <=>, ..., ??). But the requirement of Yellow is presently 5.6. Is it useful that I check all my code (when I have time :-) and rewrite in a 5.6-compatible manner? On the one hand I would say yes, on the other hand PHP 5 is already unsupported and there are tons of recommendations for upgrading the servers as soon as possible.

**Reply by markseuffert** [reply]
As you said Datenstrom Yellow requires PHP 5.6, but it's probably sooner than later that we'll switch to PHP 7.x as minimum requirement.  Do as you think is best.

**Reply by GiovanniSalmeri** [reply]
> As you said Datenstrom Yellow requires PHP 5.6, but it's probably sooner than later that we'll switch to PHP 7.x as minimum requirement. Do as you think is best.

I have added a note in the Installation section. If someone asks for a 5.6-compatible version, I will gladly help.

**Reply by markseuffert** [reply]
Sounds good. I didn't have time to answer all  your questions yesterday, here's the second part.

About making the extension self-hosted. Yes, I would include all the necessary files with an extension and would investigate options to make smaller bundles. For example if the CSS only supports modern browsers and only uses woff font files, you might be able to cut down the ballast. I don't know if this is possible with KaTeX. Sometimes libraries come with a lot of bells and whistles, additional features just in case someone needs them. When you include library files with an extension instead of using them from a CDN, you get an exact picture of the ballast and what can be done to reduce it.

Thank you for explaining the native syntax of AsciiMath. Yesterday I looked around how different Markdown implementations integrate mathematical expressions. [Math for Markdown is a hard question](https://gist.github.com/vimtaai/99f8c89e7d3d02a362117284684baa0f#math). The backtick is already used in Markdown as a special character, it means to render enclosed text as plain text, that's why I feels it's confusing when it means to do the diametrically opposite in other places. Of course we could introduce more special characters or combine special characters for new meanings. However these alternatives are a [painful experience](https://www.youtube.com/watch?v=TVwKjGbz60k) and escaping them will be too. I think adding more and more special characters to Markdown are not a good idea. This was the reason why we introduced shortcuts. For example `[image photo.jpg]` or  `[math a^2+b^2]`. I think you can intuitively see what this does and it's reasonable short to write.

**Reply by GiovanniSalmeri** [reply]
> Yes, I would include all the necessary files with an extension and would investigate options to make smaller bundles

Good, I will do so.

> that's why I feels it's confusing when it means to do the diametrically opposite in other places

I understand your point and I agree that shortcuts are easy to write and clear in their purpose, I think that I will suppress that option (or leave it undocumented). Just as a side-note, rather philosophical: one could argue that the backtick marks something as "code" (without specifying whether computer code, mathematical code, chemical code, musical code etc.), i.e. something that is not current language and that could/should be properly formatted for easier reading: syntax-coloured (for computer code), knuth-formatted (for mathematical code) and so on. In code blocks this can be signalled with an attribute, in inline code this is impossible (Pandoc markdown allows to add an inline `{.attribute}`, but this is anyway uglish). One could ask: which is the default attribute of inline code? So in my head the meaning of that option was: "choose another default meaning among the several meanings of code". But, I repeat, now I agree with you that this can be seen otherwise, and anyway changing defaults is rarely a good idea. And I perfectly agree with you regarding new special characters: much better a generic and transparent means like shortcuts. I feel that current specifications of Markdown are already *way* too complicated (see the never-ending saga of Commonmark).

**Reply by markseuffert** [reply]
Thank you for the detailed explanation, your side-note gave a very good perspective. With these information I would be in favour of removing the `MathPlainCode` functionality. I looked at the source code once more. It should be possible to remove the requirement to put mathematical expressions with space in quotes, I think the code can now directly use `$text` from `onParseContentShortcut()`. Perhaps a chance to make things simpler. 😄

**Reply by GiovanniSalmeri** [reply]
> With these information I would be in favour of removing the MathPlainCode functionality.

Removed! (In fact undocumented.)

> It should be possible to remove the requirement to put mathematical expressions with space in quotes

It's true. But on second thought I'm afraid that this would add a certain cognitive load upon users: quotes would be not only not mandatory in case of spaces, but even forbidden, and this would be different from (almost) all other shortcuts (perhaps [edit] is the only other case?). In general I see that uniform rules are more easily learnt, above all by novices.

As soon as possible I will prepare the bundle of js+css+fonts.

**Reply by GiovanniSalmeri** [reply]
> Include third party JavaScript and CSS bundles, would make the extension self-hosted.

Done 😊

**Reply by markseuffert** [reply]
Thank you, Giovanni. 

As a technical side-note, in most extensions I decided to provide only one file format for fonts. Not woff, woff2 and other file formats in parallel. There is always a tradeoff, a balance between multiple design goals. Should you care more about minimising client-side resources -vs- simplifying development tool chain -vs- using the latest technology? [There's no good answer](https://stackoverflow.com/questions/11002820/why-should-we-include-ttf-eot-woff-svg-in-a-font-face). My rule of thumb for 2021: Use only [woff](https://caniuse.com/?search=woff) if you need legacy support, otherwise only woff2. Don't know if this is something that makes sense for the Math extension or if it's possible to configure in KaTeX.

In any case, thanks again for making a Math extension. Glad you share your work with us. 👍

**Reply by GiovanniSalmeri** [reply]
> Don't know if this is something that makes sense for the Math extension or if it's possible to configure in KaTeX.

Yes, it makes perfectly sense and it is possible to configure it (even if I have chosen a rather raw form of configuration, i.e. a global substitution in css file 😊). Updated!

When I have time I would like to investigate the possibility of blending better mathematical expressions with the other text by using as far as possible the standard font of the text (the ubiquitous Computer Modern Font of Knuth seems a bit out of place...). But I'm afraid it is not very easy, mathematics use a huge quantity of special symbols and the result could be aesthetically even worse.

**Reply by zqyr** [reply]
Been waiting for this for quite some time. Gonna test it out and give you feedback - should I find any bugs.

**Reply by GiovanniSalmeri** [reply]
> Gonna test it out and give you feedback - should I find any bugs.

@zqyr Thank you very much! At least one bug is upstream: https://github.com/christianp/asciimath2tex/issues/6 Fortunately it appears quite seldom, but it is there.

**Reply by zqyr** [reply]
Upon installation, I received the following error,

`PHP Parse error:  syntax error, unexpected '>' in /home/<username>/domains/<domain>/public_html/blog/system/extensions/math.php on line 1280 <br/> `

Not sure how [this](https://github.com/GiovanniSalmeri/yellow-math/blob/d9fedc7fe3f2245dcfd07aeb503c7e3e72bd4b37/math.php#L1280) crashed my whole site.

**Reply by zqyr** [reply]
> Upon installation, I received the following error,
> 
> `PHP Parse error: syntax error, unexpected '>' in /home/<username>/domains/<domain>/public_html/blog/system/extensions/math.php on line 1280 <br/> `
> 
> Not sure how [this](https://github.com/GiovanniSalmeri/yellow-math/blob/d9fedc7fe3f2245dcfd07aeb503c7e3e72bd4b37/math.php#L1280) crashed my whole site.

**update**: Turns out I was running PHP version < 7. Anyone who comes across this issue - there: update your PHP version.

**Reply by markseuffert** [reply]
I wonder why the comparison function uses the spaceship operator, in essence it does:
`return strlen($b["asciimath"]) <=> strlen($a["asciimath"]);`

Perhaps the comparison function could use an arithmetic operation instead, for example:
`return strlen($b["asciimath"]) - strlen($a["asciimath"]);`

According to the [PHP documentation](https://www.php.net/manual/en/function.usort.php), `usort()` does not require the exact integer values of  `-1`, `0`, `+1`, it expects an integer less than, equal to, or greater than zero. I am not familiar with the inner workings of PHP, but this is how most sort functions in the C/C++ standard library work. Maybe something to try out?

**Reply by GiovanniSalmeri** [reply]
I have updated the extension: PHP >= 7 is no more required. I used the spaceship operator for no particular reason, only because it is semantically clearer (at least for my eyes) than a subtraction, and because it reminds me of Star Wars 😄. Same thing for the `??` operator, now rewritten with a ternary operator. I tested with an online syntax checker and all seems to work.

> Not sure how this crashed my whole site.

I am afraid this is the consequence of any syntax error (where error can also mean PHP incompatibilities) in any extension...

**Reply by zqyr** [reply]
Thanks. Another hiccup: upon installing math extension, and appending the following to a published post,

`[math x=(-b+-sqrt(b^2-4ac))/(2a)]`

Yellow spits out the following,

`x = \frac{- b \pm \sqrt{b^{2} - 4 a c}}{2 a}`

So yes, in essence your extension does seem to parse into AsciiMath Syntax. Am I missing anything within yellow-system.ini to render LaTEX expressions within post body?

**Reply by markseuffert** [reply]
Update your website, you need the latest extensions. Then AsciiMath should be converted into:

![image](https://user-images.githubusercontent.com/2429165/147583364-90905b1d-18ad-4c36-b781-b4f737f3f39d.png)

**Reply by GiovanniSalmeri** [reply]
A subtle bug has been corrected upstream by the author of the library and I backported the fix. Now there aren't any known issues open in asciimath2tex.

**Comment by markseuffert** [comment]
I changed my opinion about AsciiMath. I liked it as an input format in the beginning, but I currently don’t expect many people using it. As we discussed via email, Tex should be the input format for for inline `[math]` and block ` ```math `, because that’s what most people want and that's what GitHub supports in the meantime. The needs of the many outweigh the needs of the few.

**Reply by GiovanniSalmeri** [reply]
As discussed, AsciiMath support will be soon dropped from this extension: only TeX will be supported. If and when time allows, a new(ish) extension (Asciimath or Simplemath or something similar) will support only AsciiMath for the few (like me 😊) who like it.

Since this change will be anyway breaking, just a little doubt about the coming Math: a possibility, for clarity's sake, is to rename it Tex. In this way the [original suggestion](https://github.com/datenstrom/yellow/discussions/721#discussioncomment-1793904) of using `[tex]` and `` ```tex `` would be followed. Some [misunderstandings](https://en.wikipedia.org/wiki/Tex_Willer) possible only on the part of Italian users. Any opinion?

**Comment by GiovanniSalmeri** [comment]
The [Math](https://github.com/GiovanniSalmeri/yellow-math) extension has been (finally...) updated. AsciiMath support has been dropped and the [new extension Mu](https://github.com/datenstrom/yellow/discussions/975) supports only AsciiMath. As this is a breaking change, here is a quick reminder for those who are using the old Math extension:

* If you use AsciiMath, install [Mu](https://github.com/GiovanniSalmeri/yellow-mu) and rename in your pages `[math ` and `` ``` math`` to `[mu ` and `` ``` mu``
* If you use TeX/LaTeX, after installing the new version of Math rename in your pages `[mathtex ` and `` ``` mathtex`` to `[math ` and `` ``` math``

Any comments appreciated!

**Reply by markseuffert** [reply]
That's good news.

I appreciate that you took the time and splitted it into two extensions. A question about the syntax, again, if the `[math` shortcut has only one argument, would it make sense to drop the requirement to wrap input with spaces in quotes? For example then you could write something like `[math \Delta = b^2 - 4ac]`. A comment about the documentation, it looks like there's a typo or an intentional easter egg in the [screenshot](https://github.com/GiovanniSalmeri/yellow-math/blob/main/SCREENSHOT.png), If the discriminant is zero then the equation has only one distinct solution. 🤔 

Thank you for the update. I look forward to yellow-math becoming a public available extension.

**Reply by GiovanniSalmeri** [reply]
Sorry for the delay in answering 😞...

Regarding syntax, I have two little doubts about dropping the requirement to wrap input with spaces in quotes:

1. Is it indeed a simplification, or could a newbye be confused by an extension that behaves differently from most others?
3. I am working at a modification which requires an optional *second* argument, a label which allows numeration and cross-referencing. I am pretty sure that this modification will land in [Mu](https://github.com/GiovanniSalmeri/yellow-mu/), which I will use as a playground 😊. Is it useful to leave the road open for evaluating the appropriateness of this modification also in Math?

If the answer to both questions is negative, no problem in modifying the syntax!

Regarding the screenshot: actually I lazily copied the example from https://latexguide.org/book-contents/chapter-09-writing-math-formulas/, but I too was a little perplexed 🤔 Later in [Wikipedia](https://en.wikipedia.org/wiki/Quadratic_equation) I read: "If there is only one solution, one says that it is a double root". In Italian we use yet another terminology. Any suggestion to fix or clarify the text (or any suggestion of another example) is welcomed...

**Reply by markseuffert** [reply]
I think this is perplexing because a "[double root](https://math.stackexchange.com/questions/3923451/understanding-the-term-double-roots-of-a-quadratic-equation)" is not a compound word in the English language. In Sweden there is a different mathematical tradition. The quadratic equation is described as always having two solutions in the real or complex numbers. Two identical real numbers, two different real numbers, two complex numbers. Keep the easter egg. 😄

**Reply by markseuffert** [reply]
Thank you very much for mentioning the Italian terminology. I was curious what you meant. This weekend I read the Italian Wikipedia on mathematics, translated into English, a pleasure to read. What struck me often as positive was the historical context and the clear language. Such a joy to read. For example see [quadratic equation](https://it-m-wikipedia-org.translate.goog/wiki/Equazione_di_secondo_grado?_x_tr_sl=auto&_x_tr_tl=en) and [M.C. Escher](https://it-m-wikipedia-org.translate.goog/wiki/Maurits_Cornelis_Escher?_x_tr_sl=auto&_x_tr_tl=en).

**Reply by GiovanniSalmeri** [reply]
As foretold, I have updated the experimental extension [Mu](https://github.com/GiovanniSalmeri/yellow-mu). Now it has a very simple mechanism for numbering and cross-referencing expressions. I added it because numbering and cross-referencing mathematical expressions is the rule, not the exception (even though there are among experts different opinions about which expressions should be numbered [1](https://tex.stackexchange.com/questions/29267/which-equations-should-be-numbered) [2](https://academia.stackexchange.com/questions/102572/deciding-on-which-equations-to-number) [3](https://math.stackexchange.com/questions/2362196/when-should-an-equation-be-numbered-when-writing-a-paper)).

About Math, I see now three different options.

A. Relying on LaTeX's own numbering mechanism, which works out of the box in KaTeX. Well, *almost* out of the box: cross-referencing is presently supported only with a [third-party script](https://bourne2learn.com/math/katex/katex-label-tag-eqref-macro.php), which should be merged with `math.js` and tested. I see three drawbacks: 1. the LaTeX syntax is cumbersome (one should write `[math \begin{equation}\label{quad}ax^2+bx+c=0\end{equation}]` only for numbering an espression, and `[math \eqref{quad}]` for referencing it; 2. the above script adds a new constraint ("For numbering to work correctly ... each line in a numbered environment either needs to contain a \label or a \nonumber"). 3. references will be composed in math font, which typographically makes little sense.

B. Adding the same mechanism now implemented in Mu. One could simply write `[math ax^2+bx+c=0 #quad]` for numbering an expression and `[#quad]` for referencing it. I have tried this in a modified Math and it works well.

C. Neither A nor B, leaving all as it is. It is already possible to number expressions with LaTeX syntax, cross-referencing can be done by hand (but adding a link between reference and expression would be very awkward). And those who wish a handier method will switch to Mu 😁

Any opinions on which one is best?

> What struck me often as positive was the historical context and the clear language. Such a joy to read.

OT: My father would have been very glad to hear this! Mathematics was his passion and in his youth he published many papers in Italian mathematical journals (now none exists anymore). He was very proud when he discovered that Donald Knuth had cited a work of his (in Italian!) in [an article about mathematical notation](https://euclide-scuola.org/media/filelist/asalmeri/3.%20Aritmetica%20e%20Teoria%20dei%20numeri/1962%20-1992%20-%20Donald%20E.%20Knuth%2C%20Two%20notes%20on%20Notations.pdf). When he passed away I discovered a copybook with a handwritten [little treatise](https://euclide-scuola.org/media/filelist/asalmeri/3.%20Aritmetica%20e%20Teoria%20dei%20numeri/1960%20-%20Antonio%20Salmeri%2C%20Introduzione%20allo%20studio%20delle%20funzioni%20quasi-algebriche%20%28inedito%29.pdf) on "quasi-algebraic curves" (an invention of his), which later he only [partially published](https://euclide-scuola.org/media/filelist/asalmeri/3.%20Aritmetica%20e%20Teoria%20dei%20numeri/1963%20-%20Antonio%20Salmeri%2C%20Introduzione%20allo%20studio%20delle%20curve%20quasi-algebriche%2C%20GMB.pdf). Even if he wasn't particularly interested in informatics, my passion is surely due to the mathematics which I saw around me.

**Reply by markseuffert** [reply]
My advice for shortcuts would be to do as little as possible.

The following Markdown syntax might work for your extension. You can have inline shortcuts without support for cross-references and numbering, perhaps see `[math \Delta = b^2 - 4ac]` as a special form of text formatting for mathematical symbols. You can have code block shortcuts with support for cross-references and numbering, code blocks allow [special attributes](https://michelf.ca/projects/php-markdown/extra/#spe-attr), which could be used to activate numbering. About arguments without quotes being confusing for new users. It's the first thing new users will encounter, for example `[edit - You can edit this page]`. The additional hyphen bothers me, it was introduced for readability as far as I know and I have been thinking about removing it for a long time. I understand the desire for symmetry/beauty/consistency. Fortunately we are talking about mathematics and not about particle physics, perhaps see inline shortcuts with one argument as a special form. One that is useful. At the end of the day choose the Markdown syntax that fits best into your workflow.

Hope this helps.

**Reply by GiovanniSalmeri** [reply]
> see inline shortcuts with one argument as a special form

You found a good argument to convince me: I dislike *exceptions*, but *special forms* are fine 😁. I have accordingly updated the extension: syntax simplified, cross-referencing added, custom escape dropped (now closing brackets are `&rbrack;` or `\rbrack`), a bit of CSS added, many more examples in the Readme. Now all seems in good shape 😊
